fendermint/vm/interpreter/src/chain.rs

```diff
@@ -238,48 +246,31 @@ where
                         height: p.height as u64,
                         block_hash: p.block_hash,
                     };
-                    let msg = self
-                        .gateway_caller
-                        .commit_parent_finality_msg(finality.clone())?;
-                    let (state, ret) = self
-                        .inner
-                        .deliver(state, VerifiableMessage::NotVerify(msg))
-                        .await?;
-                    if ret.is_err() {
-                        return Err(anyhow!("cannot commit parent finality"));
-                    }
 
-                    // stash validator changes
+                    let (state, _, (prev_height, prev_finality)) = self
+                        .commit_finality(state, &provider, finality.clone())
+                        .await?;
+                    tracing::debug!("commit finality parsed: prev_height {prev_height}, prev_finality: {prev_finality:?}");
 
                     // error happens if we cannot get the validator set from ipc agent after retries
-                    let validator_changes = provider.validator_changes(p.height as u64).await?;
-                    let msg = self
-                        .gateway_caller
-                        .store_validator_changes_msg(validator_changes)?;
-                    let (state, ret) = self
-                        .inner
-                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                    let validator_changes = provider
+                        .validator_changes_from(prev_height + 1, p.height as u64)
+                        .await?;
+                    let (state, _, _) = self
+                        .store_validator_changes(state, validator_changes)
                         .await?;
-                    if ret.is_err() {
-                        return Err(anyhow!("failed to store validator changes"));
-                    }
-
-                    // Execute top down messages
 
-                    // error happens if we cannot get the validator set from ipc agent after retries
+                    // error happens if we cannot get the cross messages from ipc agent after retries
                     let messages = provider
-                        .top_down_msgs(p.height as u64, &finality.block_hash)
+                        .top_down_msgs_from(prev_height + 1, p.height as u64, &finality.block_hash)
                         .await?;
-                    let msg = self.gateway_caller.apply_cross_messages_msg(messages)?;
-                    let (state, ret) = self
-                        .inner
-                        .deliver(state, VerifiableMessage::NotVerify(msg))
-                        .await?;
-                    if ret.is_err() {
-                        return Err(anyhow!("failed to apply cross messages"));
-                    }
+                    let (state, ret, _) = self.execute_topdown_msgs(state, messages).await?;
```

Comment:
```suggestion
                    let (state, ret, ()) = self.execute_topdown_msgs(state, messages).await?;
```
Makes it clear what the return value is.